### PR TITLE
Add advanced deployment API

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: mingzhou.swx <mingzhou.swx@alibaba-inc.com>

### Ⅰ. Describe what this PR does
Add advanced deployment API

### CODE REVIEW

File: api/v1alpha1/deployment_types.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	// * "partitioned" means rolling in batches just like CloneSet, and will NOT create any extra Deployment;
+	// * "canary" means rolling in canary way, and will create a canary Deployment.
+	// Defaults to canary
+	DeploymentRolloutPolicyAnnotation = "rollouts.kruise.io/deployment-rolling-policy"
+)
+
+// DeploymentStrategy is strategy field for Advanced Deployment
+type DeploymentStrategy struct {
+	// original deployment strategy fields
+	apps.DeploymentStrategy
```
Comment (zmberg) [inline api/v1alpha1/deployment_types.go:27]: Is it enough to only have RollingUpdateDeployment here?

File: api/v1alpha1/deployment_types.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	// Paused = true will block the upgrade of Pods
+	Paused bool `json:"paused,omitempty"`
+	// Partition describe how many Pods should be upgraded during rollout.
+	Partition intstr.IntOrString `json:"partition,omitempty"`
+	// StableRevision record the user-specified stable revision, and we
+	// will scale down stable revision replicaset in the last.
+	StableRevision string `json:"stableRevision,omitempty"`
```
Comment (zmberg) [inline api/v1alpha1/deployment_types.go:34]: StableRevision is a bit strange in strategy

File: api/v1alpha1/deployment_types.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// ObservedGeneration record the generation of deployment this status observed.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and ready.
+	UpdatedReadyReplicas int32 `json:"updatedReadyReplicas,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and available.
+	UpdatedAvailableReplicas int32 `json:"updatedAvailableReplicas"`
+	// DesiredUpdatedReplicas is an absolute number calculated based on Partition.
+	// This field is designed to avoid users to fall into the details of algorithm
+	// for Partition calculation.
+	DesiredUpdatedReplicas int32 `json:"desiredUpdatedReplicas,omitempty"`
```
Comment (zmberg) [inline api/v1alpha1/deployment_types.go:48]: ExpectedUpdatedReplicas

File: api/v1alpha1/deployment_types.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// DeploymentExtraStatus is extra status field for Advanced Deployment
+type DeploymentExtraStatus struct {
+	// ObservedGeneration record the generation of deployment this status observed.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and ready.
+	UpdatedReadyReplicas int32 `json:"updatedReadyReplicas,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and available.
+	UpdatedAvailableReplicas int32 `json:"updatedAvailableReplicas"`
```
Comment (zmberg) [inline api/v1alpha1/deployment_types.go:44]: Refer to the definition of cloneSet, is this field unnecessary?

File: api/v1alpha1/deployment_types.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+	// DeploymentExtraStatusAnnotation is annotation for deployment,
+	// which is extra status field of Advanced Deployment.
+	DeploymentExtraStatusAnnotation = "rollouts.kruise.io/deployment-extra-status"
+
+	// DeploymentRolloutPolicyAnnotation must be "batch" or "canary":
+	// * "partitioned" means rolling in batches just like CloneSet, and will NOT create any extra Deployment;
+	// * "canary" means rolling in canary way, and will create a canary Deployment.
+	// Defaults to canary
+	DeploymentRolloutPolicyAnnotation = "rollouts.kruise.io/deployment-rolling-policy"
```
Comment (zmberg) [inline api/v1alpha1/deployment_types.go:21]: In addition, whether to consider merging these three into one annotations is also convenient for expansion. If there are only other fields in the future, there is no need to add additional annotations?

File: api/v1alpha1/deployment_types.go (hunk 1 above)
Comment (veophi) [reply api/v1alpha1/deployment_types.go:27]: done

File: api/v1alpha1/deployment_types.go (hunk 3 above)
Comment (veophi) [reply api/v1alpha1/deployment_types.go:48]: done

File: api/v1alpha1/deployment_types.go (hunk 4 above)
Comment (veophi) [reply api/v1alpha1/deployment_types.go:44]: deleted

File: api/v1alpha1/deployment_types.go (hunk 5 above)
Comment (veophi) [reply api/v1alpha1/deployment_types.go:21]: DeploymentRolloutPolicyAnnotation can be merged into strategy annotation, but extra status anno should be independent.

File: api/v1alpha1/deployment_types.go (hunk 1 above)
Comment (furykerry) [inline api/v1alpha1/deployment_types.go:27]: is it necessary to duplicate apps.DeploymentStrategy here ? can we use the deploymentStrategy in the deployment spec?

File: api/v1alpha1/deployment_types.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	UpdatedReadyReplicas int32 `json:"updatedReadyReplicas,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and available.
+	UpdatedAvailableReplicas int32 `json:"updatedAvailableReplicas"`
+	// DesiredUpdatedReplicas is an absolute number calculated based on Partition.
+	// This field is designed to avoid users to fall into the details of algorithm
+	// for Partition calculation.
+	DesiredUpdatedReplicas int32 `json:"desiredUpdatedReplicas,omitempty"`
+}
+
+func SetDefaultDeploymentStrategy(strategy *DeploymentStrategy) {
```
Comment (furykerry) [inline api/v1alpha1/deployment_types.go:51]: if we reuse MaxSurge and MaxUnavailable in the deployment spec, there is no need to set defaults

File: pkg/controller/deployment/util/deployment_util.go [hunk 7]
```diff
@@ -894,3 +897,28 @@ func (o ReplicaSetsBySizeNewer) Less(i, j int) bool {
...
+		return false
+	}
+	if deployment.Spec.Strategy.Type != apps.RecreateDeploymentStrategyType {
+		return false
+	}
+	return deployment.Spec.Paused
+}
+
+// NewRSReplicasLimit return a limited replicas of new RS calculated via partition.
+func NewRSReplicasLimit(partition intstrutil.IntOrString, deployment *apps.Deployment) int32 {
```
Comment (furykerry) [inline pkg/controller/deployment/util/deployment_util.go:916]: plz add ut

File: pkg/controller/deployment/controller.go [hunk 8]
```diff
@@ -167,26 +174,49 @@ func (r *ReconcileDeployment) Reconcile(_ context.Context, request reconcile.Req
...
 // TODO: create new controller only when deployment is under our control
-func (f *controllerFactory) NewController(_ *appsv1.Deployment) (*DeploymentController, error) {
+func (f *controllerFactory) NewController(deployment *appsv1.Deployment) *DeploymentController {
+	if !deploymentutil.IsUnderRolloutControl(deployment) {
+		klog.Warningf("Deployment %v is not under rollout control, ignore", klog.KObj(deployment))
+		return nil
+	}
+
+	strategy := rolloutsv1alpha1.DeploymentStrategy{}
+	strategyAnno := deployment.Annotations[rolloutsv1alpha1.DeploymentStrategyAnnotation]
```
Comment (zmberg) [inline pkg/controller/deployment/controller.go:197]: Who inject the annotations first, webhook?

File: pkg/controller/deployment/util/deployment_util.go (hunk 7 above)
Comment (veophi) [reply pkg/controller/deployment/util/deployment_util.go:912]: done

**Pull request merged**